Add IO expend IC CAT9554 support

SchumyHao opened the pull request:
## Description:

Add IO expend IC CAT9554 support.

**Related issue (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** https://github.com/esphome/esphome-docs/pull/448

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

SchumyHao committed:
Add cat 9554 support

Signed-off-by: SchumyHao <schumyhaojl@126.com>

SchumyHao committed:
Fix build errors

Signed-off-by: SchumyHao <schumyhaojl@126.com>

Commit:
Add B/W support for Waveshare 2.90in (B) screen (#889)

* Add B/W support for Waveshare 2.90in (B) screen

* Fix for() loop source code formatting

glmnet committed:
fix climate_ir on receive optional (#897)

* fix climate on receive optional

* add climate tests

valordk committed:
Add support for Sensirion SPS30 Particulate Matter sensors (#891)

* Add support for Sensirion SPS30 Particulate Matter sensors

* Remove blocking of the main thread on initialization;
Improve wording on the debug messages;
Add robustness in re-initialization of reconnected or replaced sensors;

* Fix code formatting;


Co-authored-by: Nad <valordk@github>

freekode committed:
Add TM1561 support (#893)

* Add TM1561 support

* Fixed after clang-tidy

* Fixed after clang-tidy

* Fixed after clang-tidy, updated lib_deps

* Fixed after clang-tidy, updated formatting

* Added actions, removed from display domain

* Protected methods naming

* float casting

* float casting

OttoWinter committed:
Add magic value REPLACEME (#881)

* Add magic value REPLACEME

* Lint

OttoWinter committed:
Fix MQTT logs Int or String expected Python 3 (#898)

Fixes https://github.com/esphome/issues/issues/850

OttoWinter committed:
web_server call setup_controller (#899)

OttoWinter committed:
Pulse counter validate not both disabled (#902)

OttoWinter committed:
Optimize application loop speed (#860)

* Optimize application loop speed

* Also check call_loop

* Remove duplicate code

* Fixes

OttoWinter committed:
Ignore ESP32 Camera unknown framesizes (#901)

Fixes https://github.com/esphome/issues/issues/866

OttoWinter committed:
Better/stricter pin validation (#903)

* Better/stricter pin validation

* Update tests

OttoWinter committed:
Disable default wait_time for rc_switch (#900)

See also https://github.com/esphome/esphome/commit/82625a30808f4c914bb9e6a2b3e7ef229a451c99#commitcomment-36092052

OttoWinter committed:
Update ESP32 BLE ADV parse to match BLE spec (#904)

* Update ESP32 BLE ADV parse to match BLE spec

* Update xiaomi

* Update ruuvi

* Format

* Update esp32_ble_tracker.cpp

* Fix log

* Format

* Update xiaomi_ble.cpp

OttoWinter committed:
Update dependencies (#906)

PyYAML 5.1.2 -> 5.2 https://github.com/yaml/pyyaml/blob/master/CHANGES
flake8 3.6.0 -> 3.7.9 https://github.com/PyCQA/flake8/tree/master/docs/source/release-notes
paho-mqtt 1.4.0 -> 1.5.0 https://github.com/eclipse/paho.mqtt.python/blob/master/ChangeLog.txt
platformio 4.0.3 -> 4.1.0 https://github.com/platformio/platformio-core/releases
protobuf 3.10.0 -> 3.11.1 https://github.com/protocolbuffers/protobuf/releases
pylint 2.3.0 -> 2.4.4 http://pylint.pycqa.org/en/latest/whatsnew/changelog.html#what-s-new-in-pylint-2-4-4

OttoWinter committed:
Add ESP8266 core v2.6.2 (#905)

* Add ESP8266 core v2.6.2

* Upstream ESP8266 Wifi fixes

* Replace disable_interrupt with InterruptLock C++ class

* Update code to use InterruptLock

* Lint

* Update dht.cpp

* Improve InterruptLock docs, mark as ICACHE_RAM_ATTR

* Fixes

OttoWinter committed:
Fix ESP32 interrupt enable/disable switched

Needs to be manually cherry-picked

edge90 committed:
Handle yaml merge keys correcly. (#888)

* Handle yaml merge keys correcly.

* Removed old debug bool.

* Deleted after request from @OttoWinder.

* Small refactoring.

Removed unused variable `value`
Small refactoring to make the code clearer.
Added comments.

* Fix merge sequence edge case

jesserockz committed:
Allow loading esphome version from a fork (#907)

SchumyHao committed:
Fix travis fail issue

Signed-off-by: SchumyHao <schumyhaojl@126.com>

SchumyHao committed:
Fix some format issue

Signed-off-by: SchumyHao <schumyhaojl@126.com>

OttoWinter committed:
Clean up YAML Mapping construction (#910)

* Clean up YAML Mapping construction

Fixes https://github.com/esphome/issues/issues/902

* Clean up DataBase

* Update error messages

OttoWinter commented on esphome/components/cat9554/__init__.py:
```diff
@@ -0,0 +1,61 @@
+CONF_IRQ_PIN = 'irq_pin'
+CONFIG_SCHEMA = cv.Schema({
+    cv.Required(CONF_ID): cv.declare_id(CAT9554Component),
+    cv.Required(CONF_IRQ_PIN): pins.gpio_input_pin_schema,
```
If an interrupt is registered on this pin, you should validate it using `pins.internal_gpio_input_pin_schema` (interrupts can't be registered on pins that are not on the ESP itself).

OttoWinter commented on esphome/components/cat9554/__init__.py:
```diff
@@ -0,0 +1,61 @@
+    value = cv.string(value)
+    if value.upper() == 'INPUT_PULLUP':
+        raise cv.Invalid("INPUT_PULLUP mode has been removed in 1.14 and been combined into "
+                         "INPUT mode (they were the same thing). Please use INPUT instead.")
```
This was only for the PCF8574 integration (for which input_pullup mode was removed in 1.14). Since cat9554 never had this mode, this warning+check can be removed.

So replace each instance of `validate_cat9554_gpio_mode` with `cv.enum(CAT9554_GPIO_MODES, upper=True)`.

OttoWinter commented on esphome/components/cat9554/__init__.py:
```diff
@@ -0,0 +1,61 @@
+
+CAT9554_OUTPUT_PIN_SCHEMA = cv.Schema({
+    cv.Required(CONF_CAT9554): cv.use_id(CAT9554Component),
+    cv.Required(CONF_NUMBER): cv.int_,
```
How many pins does the cat9554 have? It would be good to validate it here using `cv.int_range()` (see definition for arguments)

OttoWinter commented on esphome/components/cat9554/cat9554.cpp:
```diff
@@ -0,0 +1,140 @@
+
+static const char *TAG = "cat9554";
+
+static CAT9554Component *instance_;
```
What is this global variable used for? you can pass `this` directly to the `attach_interrupt` call.

OttoWinter commented on esphome/components/cat9554/cat9554.cpp:
```diff
@@ -0,0 +1,140 @@
+
+void CAT9554Component::setup() {
+  ESP_LOGCONFIG(TAG, "Setting up CAT9554...");
+  ESP_LOGCONFIG(TAG, "    Address: 0x%02X", this->address_);
```
This line should be removed - newer esphome versions dump the config in `dump_config`, not in setup()

OttoWinter commented on esphome/components/cat9554/cat9554.cpp:
```diff
@@ -0,0 +1,140 @@
+  instance_ = this;
+  this->irq_pin_->setup();
+  this->isr_ = this->irq_pin_->to_isr();
+  this->irq_pin_->attach_interrupt(gpio_intr, &instance_, FALLING);
```
This interrupt code will randomly crash the ESP. The reasons are complicated, but basically the flash cannot be accessed while an ISR is running and all functions called from the ISR must be tagged `ICACHE_RAM_ATTR`. 

However, `update_gpio_needed` is not placed in instruction cache *plus* since `CAT9554Component` has virtual members, the vtable must be accessed to calculate the offset from the pointer. And vtables are stored in flash which means you have to use a struct to communicate without vtables.

I don't really see a reason why the IRQ line is used here. Polling the CAT9554 in each loop is fine (as demonstrated by pcf8574 which this is based on).

I would suggest removing IRQ and having the pin be polled instead.

SchumyHao replied:
You means polling this pin state? Or polling extend IC GPIOs status though I2C?
If change irq to polling pin state, this maybe acceptable, but polling extend IC GPIOs status though I2C is not efficient.

OttoWinter replied:
Polling the IC.

Yes it's not 100% efficient - but on the other hand the main esphome loop is running at 60Hz no matter what. Reading an IC won't change that (as can be seen with PCF8574 and other I/O expander integrations).

OttoWinter commented on esphome/components/cat9554/cat9554.cpp:
```diff
@@ -0,0 +1,140 @@
+  if (this->is_failed())
+    return false;
+
+  if (!this->write_byte(INPUT_REG & 0xff, this->config_mask_)) {
```
```suggestion
  if (!this->write_byte(INPUT_REG, this->config_mask_)) {
```

Why all these `& 0xFF`? It should be clear from the enum definition that these are all values that fit into 8-bit integers.

nickw444 committed:
Add duty cycle output component (#894)

* Add duty cycle output component

* cleanup + tests

* format

* duty_cycle -> slow_pwm

* .

* clang-format

* ESP_LOGD -> ESPLOGVV

Co-Authored-By: Otto Winter <otto@otto-winter.com>

adamgreg committed:
ESP32 GPIOs 33 to 38 can be used for deep sleep wakeup (#911)

OttoWinter committed:
Drop Python 2 Support (#793)

* Remove Python 2 support

* Remove u-strings

* Remove docker symlinks

* Remove from travis

* Update requirements

* Upgrade flake8/pylint

* Fixes

* Manual

* Run pyupgrade

* Lint

* Remove base_int

* Fix

* Update platformio_api.py

* Update component.cpp

OttoWinter committed:
Fix CI

jesserockz committed:
Add RFBridge component (#896)

* Add RFBridge component

* Fix format issues

* Rename methods

* More formatting

* Fix line length

* Apply suggestions from code review

Co-Authored-By: Otto Winter <otto@otto-winter.com>

* Check uart settings on dump

* Make receiving local to the loop

* FIx code order and schema

* Add rf_bridge to test file

* Apply suggestions from code review

Co-Authored-By: Otto Winter <otto@otto-winter.com>

SchumyHao committed:
Accepte some advice from ottowinter. Change IRQ pin as optional

Signed-off-by: SchumyHao <schumyhaojl@126.com>

SchumyHao committed:
accept some advice from ottowinter

Signed-off-by: SchumyHao <schumyhaojl@126.com>

SchumyHao committed:
Make irq pin as a optional pin

Signed-off-by: SchumyHao <schumyhaojl@126.com>

SchumyHao committed:
Fix clang-format

Signed-off-by: SchumyHao <schumyhaojl@126.com>

balrog-kun committed:
ct_clamp: Check sample() return value is not NaN (#921)

Don't try to update CT clamp's state with NaN values returned from the
underlaying sensor.  A single IO error in the sensor code will cause a
NaN to be returned and if we use that in CTClampSensor's floating point
maths both sample_sum_ and offset_ will become NaN and from there every
future calculation will use the NaN offset_ and return NaN too.

SchumyHao commented:
Hi @OttoWinter   Can you tell me how to generate svg logo for CAT9554? 

There is no pretty image for this IC, I need use svg just show CAT9554

SchumyHao committed:
Add test config for cat9554

Signed-off-by: SchumyHao <schumyhaojl@126.com>

glmnet committed:
Climate Mitsubishi (#725)

* add climate

* Mitsubishi updates

* refactor mitsubishi to use climate_ir

* lint

wilmardo committed:
fix: only decode when not str already (#923)

Signed-off-by: wilmardo <info@wilmardenouden.nl>

glmnet committed:
fix climate-ir bad merge (#935)

* fix climate-ir bad merge

* add mitshubishi test

Anonym-tsk committed:
http_request: fix memory allocation (#916)

Anonym-tsk committed:
http_request version fix (#917)

OttoWinter committed:
PID Climate (#885)

* PID Climate

* Add sensor for debugging PID output value

* Add dump_config, use percent

* Add more observable values

* Update

* Set target temperature

* Add autotuner

* Add algorithm explanation

* Add autotuner action, update controller

* Add simulator

* Format

* Change defaults

* Updates

timsavage committed:
Use b''.decode() instead of str(b'') (#941)

Handling of request arguments in WizardRequestHandler is not decoding
bytes and rather just doing a str conversion resulting in a value of
"b''" being supplied to the wizard code.

Valcob committed:
Adding the espressif 2.6.3 (#944)

Commit:
extract and use current version of python 3 (#938)

voibit committed:
Inverted output in neopixelbus (#895)

* Added inverted output

* Added support for inverted output in neopixelbus

* Update esphome/components/neopixelbus/light.py

Co-Authored-By: Otto Winter <otto@otto-winter.com>

* Update light.py

* corrected lint errors

Co-authored-by: Otto Winter <otto@otto-winter.com>

cyberplant committed:
Added degree symbol for MAX7219 7-segment display. (#764)

The ascii char to use it is "~" (0x7E).

Disclaimer: I didn't test this yet.

andrasbiro committed:
Fix dump/tx of 64 bit codes (#940)

* Fix dump/tx of 64 bit codes

* fixed source format

dmkif committed:
Update hdc1080.cpp (#887)

* Update hdc1080.cpp

increase waittime, to fix reading errors

* Fix: Update HDC1080.cpp

i fixed the my change on write_bytes

glmnet committed:
add tcl112 support for dry, fan and swing (#939)

panuruj committed:
Fix SGP30 incorrect baseline reading/writing (#936)

* Split the SGP30 baseline into 2 values

- According to the SGP30 datasheet, each eCO2 and TVOC baseline is a 2-byte value (MSB first)
- The current implementation ignores the MSB of each of the value
- Update the schema to allow 2 different baseline values (optional, but both need to be specified for the baseline to apply)

* Make both eCO2 and TVOC required if the optional baseline is defined

* Make dump_config() looks better

ayufan committed:
Add register_*_effect to allow registering custom effects (#947)

This allows to register custom effect from user components,
allowing for bigger composability of source.

timsavage committed:
Bugfix/normalize core comparisons (and Python 3 update fixes) (#952)

* Correct implementation of comparisons to be Pythonic

If a comparison cannot be made return NotImplemented, this allows the
Python interpreter to try other comparisons (eg __ieq__) and either
return False (in the case of __eq__) or raise a TypeError
exception (eg in the case of __lt__).

* Python 3 updates

* Add a more helpful message in exception if platform is not defined

* Added a basic pre-commit check

kbx81 committed:
Add transmit pioneer (#922)

* Added pioneer_protocol to support transmit_pioneer

glmnet committed:
Display tm1637 (#946)

* add TM1637 support

mario-tux committed:
Support a further variant of Xiaomi CGG1 (#930)

hectorgimenez committed:
Daikin climate ir component (#964)

* Daikin ARC43XXX IR remote controller support

* Format and lint fixes

* Check temperature values against allowed min/max

glmnet committed:
fix tm1637 missing __init__.py (#975)

andriej committed:
sim800l: Add support of roaming-registered SIM cards (#977)

* Add support of roaming-registered cards

* Change or to ||

gmta committed:
BME280: fix typos, use forced mode constant (#974)

* Fix typo in BME280 chip ID error message

* Use BME280 forced mode constant instead of literal

puuu committed:
MQTT climate features (#913)

* mqtt_climate: add action support

* mqtt_climate: add fan and swing mode support

* mqtt_climate: reduce length of discovery payload by using abbreviations

https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/mqtt/abbreviations.py

OttoWinter committed:
Revert ESP32 default upload speed to 115200 (#978)

freekode committed:
Add TM1651 simple level, turn on, turn off actions (#920)

* Add TM1651 simple level action

* fixed brightness validation

* Updated lib, fixed import

* Added turn_on, turn_off actions

* Fixed after lint

Elkropac committed:
Webserver - include css, js in index (#932)

* add new config options

* init variables in code

* load css and js in python

* update print inside webserver

* fix indentation

* fix indentation

* indentation fix

* fix condition in init

* use cv.file_ instead of cv.string

* do not import EsphomeError

* support embedding js and css at the same time as defined in url

* handle css as separate page

* handle js as separate page

* fix copy and paste error

egeltje committed:
Added equal symbol for MAX7219 7-segment display (#986)

brandond committed:
Output from platformio click command does not need to be decoded (#953)

brandond committed:
Allow custom lights to be addressable (#954)

pauln committed:
Fix esphome/issues#947 - RGBW(W) white brightness (#925)

thomasklingbeil committed:
Add support for TTGO epaper boards with B73 revision (#928)

* Add support for TTGO epaper boards with B73 revision

Skaronator committed:
Fix OTA updates getting killed by task_wdt (#959)

timsavage committed:
Bugfix/1077 decode called on str fetching platformio stacktrace (#991)

* Remove decode from str result, add type annotations

buxtronix committed:
Add support for Tuya ceiling fan controllers (#989)

* Add support for Tuya ceiling fan controllers

sekkr1 committed:
Fixed iBeacon struct and major and minor parsing (#987)

Co-authored-by: sekkr1 <dekel.reches@pm.me>

Anonym-tsk committed:
http_request http fix (#980)

Co-authored-by: Nikolay Vasilchuk <nikolay.vasilchuk@corp.mail.ru>

quinnhosler committed:
Rgbww color fix (#967)

* RGBWW color vs white and brightness adjustments

glmnet committed:
add time cover assumed_state option (#979)

escoand committed:
add on_rc_switch trigger (#983)

Sizurka committed:
SCD30 fixes and improvements (#962)

* SCD30 improvements

yekm committed:
pzemac total energy support (#933)

* add energy support in pzemac sensor

Co-authored-by: Sergio Mayoral Martinez <sergiomayoralmartinez@gmail.com>
Co-authored-by: t151602 <sergio.mayoralmartinez@telefonica.com>
Co-authored-by: Otto Winter <otto@otto-winter.com>

Azimath committed:
Add TMP117 component (#992)

* Create TMP117 sensor component

timsavage committed:
Unittests for esphome python code (#931)

CircuitSetup committed:
Corrections to default register values of ATM90E32 component (#982)

* Corrections to default register values of ATM90E32 component

gmasse committed:
Add AHT10 sensor (#949)

abmantis committed:
wifi: retry connection if the connection is not valid (#994)

pax0r committed:
Support for pcd8544 (nokia 5110 and 3310) screen (#973)

* First version of working compontent for pc8544 screen

* Fixed lint errors

* Fixed lint errors #2

glmnet committed:
fix servo bug restoring state and starting servo detached (#1008)

Anonym-tsk committed:
VSCode devcontainer support (#914)

* Devcontainer

* Removed header from json

glmnet commented:
LGTM docs are missing an image

glmnet committed:
removes comments from lambda (#998)

* removes comments from lambda

* include comments in lambda test

* pylint no else return

freekode committed:
Add HM3301 laser dust detection sensor (#963)

* Add HM3301 laser dust detection sensor

* Fixed after lint

* Fixed after lint

* added status clear warning

kroimon committed:
Add constant_brightness property to CWWW/RGBWW lights (#1007)

Fixes https://github.com/esphome/feature-requests/issues/460

Co-authored-by: Otto Winter <otto@otto-winter.com>

balrog-kun committed:
web_server: Add cover calls to REST API (#999)

Add the GET and POST handler for cover components.  Also add covers to
the index page although the Open/Close buttons that are shown for covers
will need a few lines added to webserver-v1.js, without them they don't
do anything.

OttoWinter committed:
Add AC Dimmer support (#880)

* Add AC Dimmer support

Fixes https://github.com/esphome/feature-requests/issues/278

* fixes

basically missed the output pin setup and in the switching was switching true true true :P

* Format

* Enable ESP32

* Also setup ZC pin

* Support multiple dimmers sharing ZC pin

* Fix ESP32

* Lint

* off gate on zc detect

* tests pins validation

* Climate Mitsubishi (#725)

* add climate

* Mitsubishi updates

* refactor mitsubishi to use climate_ir

* lint

* fix: only decode when not str already (#923)

Signed-off-by: wilmardo <info@wilmardenouden.nl>

* fix climate-ir bad merge (#935)

* fix climate-ir bad merge

* add mitshubishi test

* http_request: fix memory allocation (#916)

* http_request version fix (#917)

* PID Climate (#885)

* PID Climate

* Add sensor for debugging PID output value

* Add dump_config, use percent

* Add more observable values

* Update

* Set target temperature

* Add autotuner

* Add algorithm explanation

* Add autotuner action, update controller

* Add simulator

* Format

* Change defaults

* Updates

* Use b''.decode() instead of str(b'') (#941)

Handling of request arguments in WizardRequestHandler is not decoding
bytes and rather just doing a str conversion resulting in a value of
"b''" being supplied to the wizard code.

* Adding the espressif 2.6.3 (#944)

* extract and use current version of python 3 (#938)

* Inverted output in neopixelbus (#895)

* Added inverted output

* Added support for inverted output in neopixelbus

* Update esphome/components/neopixelbus/light.py

Co-Authored-By: Otto Winter <otto@otto-winter.com>

* Update light.py

* corrected lint errors

Co-authored-by: Otto Winter <otto@otto-winter.com>

* Added degree symbol for MAX7219 7-segment display. (#764)

The ascii char to use it is "~" (0x7E).

Disclaimer: I didn't test this yet.

* Fix dump/tx of 64 bit codes (#940)

* Fix dump/tx of 64 bit codes

* fixed source format

* Update hdc1080.cpp (#887)

* Update hdc1080.cpp

increase waittime, to fix reading errors

* Fix: Update HDC1080.cpp

i fixed the my change on write_bytes

* add tcl112 support for dry, fan and swing (#939)

* Fix SGP30 incorrect baseline reading/writing (#936)

* Split the SGP30 baseline into 2 values

- According to the SGP30 datasheet, each eCO2 and TVOC baseline is a 2-byte value (MSB first)
- The current implementation ignores the MSB of each of the value
- Update the schema to allow 2 different baseline values (optional, but both need to be specified for the baseline to apply)

* Make both eCO2 and TVOC required if the optional baseline is defined

* Make dump_config() looks better

* Add register_*_effect to allow registering custom effects (#947)

This allows to register custom effect from user components,
allowing for bigger composability of source.

* Bugfix/normalize core comparisons (and Python 3 update fixes) (#952)

* Correct implementation of comparisons to be Pythonic

If a comparison cannot be made return NotImplemented, this allows the
Python interpreter to try other comparisons (eg __ieq__) and either
return False (in the case of __eq__) or raise a TypeError
exception (eg in the case of __lt__).

* Python 3 updates

* Add a more helpful message in exception if platform is not defined

* Added a basic pre-commit check

* Add transmit pioneer (#922)

* Added pioneer_protocol to support transmit_pioneer

* Display tm1637 (#946)

* add TM1637 support

* Support a further variant of Xiaomi CGG1 (#930)

* Daikin climate ir component (#964)

* Daikin ARC43XXX IR remote controller support

* Format and lint fixes

* Check temperature values against allowed min/max

* fix tm1637 missing __init__.py (#975)

* Add AC Dimmer support

Fixes https://github.com/esphome/feature-requests/issues/278

* fixes

basically missed the output pin setup and in the switching was switching true true true :P

* Format

* Enable ESP32

* Also setup ZC pin

* Support multiple dimmers sharing ZC pin

* Fix ESP32

* Lint

* off gate on zc detect

* tests pins validation

* fix esp8266 many dimmers, changed timing

* Increased value resolution, added min power

* use min_power from base class

* fix min_power. add init with half cycle

* added method for trailing pulse, trailing and leading

* fix method name. try filter invalid falling pulse

* renamed to ac_dimmer

* fix ESP32 not configuring zero cross twice

Co-authored-by: Guillermo Ruffino <glm.net@gmail.com>
Co-authored-by: Wilmar den Ouden <wilmardo@users.noreply.github.com>
Co-authored-by: Nikolay Vasilchuk <Anonym.tsk@gmail.com>
Co-authored-by: Tim Savage <tim@savage.company>
Co-authored-by: Vc <37367415+Valcob@users.noreply.github.com>
Co-authored-by: gitolicious <mrjchn@gmail.com>
Co-authored-by: voibit <krestean@gmail.com>
Co-authored-by: Luar Roji <cyberplant@users.noreply.github.com>
Co-authored-by: András Bíró <1202136+andrasbiro@users.noreply.github.com>
Co-authored-by: dmkif <dmkif@users.noreply.github.com>
Co-authored-by: Panuruj Khambanonda (PK) <pk@panurujk.com>
Co-authored-by: Kamil Trzciński <ayufan@ayufan.eu>
Co-authored-by: Keith Burzinski <kbx81x@gmail.com>
Co-authored-by: Mario <4376789+mario-tux@users.noreply.github.com>
Co-authored-by: Héctor Giménez <hector.fwbz@gmail.com>

reidprojects committed:
feat: Add support for MCP23016 IO Expander (#1012)

* feat: Add support for MCP23016 IO extander

* fix: Fix style

puuu committed:
Daikin climate receiver support (#1001)

* climate.daikin: implement remote receive

* climate.daikin: fix temperature value in special modes

* climate.daikin: tweak timing to fit better to ir-remote signal

timsavage committed:
Tests for CPP Code generation and some Python3 improvements (#961)

* Basic pytest configuration

* Added unit_test script that triggers pytest

* Changed "fixtures" to fixture_path

This is consistent with pytest's tmp_path

* Initial unit tests for esphome.helpers

* Disabled coverage reporting for esphome/components.

Focus initial unittest efforts on the core code.

* Migrated some ip_address to hypothesis

* Added a hypothesis MAC address strategy

* Initial tests for core

* Added hypothesis to requirements

* Added tests for core classes

TestTimePeriod
Lambda
ID
DocumentLocation
DocumentRange
Define
Library

* Updated test config so package root is discovered

* Setup fixtures and inital tests for pins

* Added tests for validate GPIO

* Added tests for pin type

* Added initial config_validation tests

* Added more tests for config_validation

* Added comparison unit tests

* Added repr to core.TimePeriod. Simplified identifying faults in tests

* Fixed inverted gt/lt tests

* Some tests for Espcore

* Updated syntax for Python3

* Removed usage of kwarg that isn't required

* Started writing test cases

* Started writing test cases for cpp_generator

* Additional docs and more Python3 releated improvements

* More test cases for cpp_generator.

* Fixed linter errors

* Add codegen tests to ensure file API remains stable

* Add test cases for cpp_helpers

glmnet committed:
Climate whirlpool (#1029)

* wip

* transmitter ready

* climate.whirlpool receiver implemented (#971)

* receiver implemented

* Support for two models of temp ranges

* temperature type and lint

* more lint

Co-authored-by: Guillermo Ruffino <glm.net@gmail.com>

* add test

* not mess line endings

Co-authored-by: mmanza <40872469+mmanza@users.noreply.github.com>

glmnet committed:
add mac address to wifi info (#1030)

* add mac address to wifi info

* add test

* lint

Sizurka committed:
SHTC3: Wake up the sensor during setup (#993)

Since we put the sensor to sleep when not measuring, make sure to wake
it up during setup.  It does not respond to any other I2C commands
(including reset) while asleep, so if we've done a soft reboot while
it was sleeping then setup() would fail since it wouldn't respond to
anything.

ukewea committed:
Change buffer sending process for waveshare_epaper (2.70in) (#1031)

* Change buffer sending process for waveshare_epaper (2.70in)

The current way ESPhome sending buffer to WaveshareEPaper2P7In does not show the expected content on the display, this commit is changing the data transferring process so the content is showing as expected.

The process is adapted from the demo code provided by Waveshare, manufacturer of the E-paper display.

* Fix linting eror

glmnet committed:
add lights on off triggers (#1037)

* add lights on off triggers

* add test

puuu committed:
Bluetooth advertising automation (#995)

* esp32_ble_tracker: introduce UUID comparison function

* ble_presence, ble_rssi: use new UUID comparison function

* esp32_ble_tracker: introduce automation on BLE advertising

* test2.yaml: remove deep_sleep due to firmware size restrictions

fake-name committed:
Fix missing yield in uart causing watchdog timer resets in esp32 when blocking waiting on serial responses. (#1016)

Skaronator committed:
Make initial run variable available to addressable_lambda (#1035)

* Make initial run variable available to addressable_lambda

* Fix linting

* Remove if clause

jonathanadams committed:
Dashboard Updates (#1025)

* Restructuring of static files directory

* Update jQuery to v3.5.0

* Update jQuery UI to v1.12.1

* Update jQuery Validate to v1.19.1

* Improve login page layout and color scheme

* Updated header & footer colour scheme & added ESPHome logo

* Restructuring of fonts directory

* Restructuring of fonts directory

* Corrected icon reference error

* Update node layout and styling

* Update Ace to v1.4.10

* JS file reorganisation

* Rewrite of LogModal class including refactorization of js & html

* Updated delete node function

* Rewrite of editor modal

* Update Materialize Stepper to v3.1.0

* Updates to the wizard modal

* Added wizard validators back in and removed comments

* Merge old stylesheet into new

* Linting errors

* Fixed dashboard layout issue when no nodes present

* Introduced dynamic 3 column layout for large screens

* Removed unnecessary code

* Update data attribute names

* Added loading indicator to editor

* Open validator websocket on document ready

* Automatically restart validator websocket if it closes

* Minor styling updates

* Improvements to on boarding process

* Node display filename and then path on hover

* Removing console.logs

* Minor styling revisions

* Added toast on begin

* Fix lint

Co-authored-by: Sergio Mayoral Martinez <sergiomayoralmartinez@gmail.com>

puuu committed:
esp32 remote: make RMT memory blocks configureable (#1002)

glmnet committed:
test disable no delay (#1026)

* test disable no delay

* removed set_nodelay code

Anonym-tsk committed:
http_request ESP32 insecure requests fix (#1041)

* Fixed #1175

* CI

Co-authored-by: Nikolay Vasilchuk <nikolay.vasilchuk@corp.mail.ru>

teamsuperpanda committed:
Update __init__.py (#1020)

buxtronix committed:
Some max7219 updates. (#1021)

* Some max7219 updates. Disable test mode on startup and allow dynamic brightness

* fix formatting

sredfern committed:
Extending Support to 5.83in Waveshare eink B/W displays (#1009)

* Extending Support to 5.83in Waveshare eink B/W displays

* Fixed lint issues

jesserockz committed:
Allow tm1637 to use pins from IO expanders (#1058)

* Allow tm1637 to use pins from IO expanders

* Add test for io expander pins

blejdfist committed:
Fix fan oscillation trait not being used (#1048)

nepozs committed:
Update tm1637.cpp (#1044)

* Update tm1637.cpp

Defined degree symbol.

* remove fixed array size

Co-authored-by: Guillermo Ruffino <glm.net@gmail.com>

robinsmidsrod committed:
dht: Fix sensor reading from DHT22 (#926)

* dht: Fix sensor reading from DHT22 (#239)

Looking at the datasheet for DHT22, it claims to need a delay of 18ms, not
800us. Change to use the same delay as DHT11. Tested working with NodeMCUv3
and DHT22 sensor with board with built-in resistor.

* dht: Add model DHT22_TYPE2 with 2ms delay instead of 0.8ms

The model DHT22_TYPE2 is exactly the same as DHT22, but uses a different
delay.

* dht: Fix bogus negative temperature reading on DHT22_TYPE2

The workaround for negative numbers associated with DHT22s can be skipped
on these sensors.

jesserockz committed:
Add lambda to devcontainer config (#1059)

jzlka committed:
Ble scanner (#976)

* Added ble_scanner component

* ble_scanner: time.h changed to ctime

* ble_scanner: Test added.

* ble_scanner: fixed code formatting.

* Removed duplicate binary sensor tests from test2 and test3 to decrease memory requirements.

* Removed another duplicate test from test2.yaml and moved stepper test to yaml3.yaml to reduce memory requirements of test2.yaml.

* Reverted the last stepper test change

* Moved some sensor tests from test2.yaml to test3.yaml to save memory.

* Moved ruuvitag back to test2.yaml as it requires component esp32_ble_tracker.

* removed apds9960 as it is duplicated

Co-authored-by: Guillermo Ruffino <glm.net@gmail.com>

Skaronator committed:
Upgrade dependencies

krahabb committed:
Added auto discovery and setup to Dallas Platform (#1028)

* Added auto discovery and setup to Dallas Platform

* Added auto discovery and setup to Dallas Platform

* Added auto discovery and setup to Dallas Platform

* Added auto discovery and setup to Dallas Platform

* Added auto discovery and setup to Dallas Platform

ctalkington committed:
expose mac address via discovery (#1038)

napieraj committed:
Add support for ESP32 DAC (#1071)

* Add support for ESP32 onboard DAC

* Newlines

* Tests

0hax committed:
Uart improvments (#1024)

* uart: Add support for specifying the number of bits and parity.

ESP8266SwSerial doesn't really check parity but just read the parity bit
and ignore it when receiving data.

Signed-off-by: 0hax <0hax@protonmail.com>

* uart: support begin and end methods.

A component may need to reset uart buffer/status by using begin() and
end() methods. This is useful for example when a component needs to be
sure it is not reading garbage from previously received data over uart.
For end() methods with software serial, disabling interrupt is
currently impossible because of a bug in esp8266 Core:
https://github.com/esp8266/Arduino/issues/6049

Signed-off-by: 0hax <0hax@protonmail.com>

* esphal: add support for detaching an interrupt.

That's needed when a component needs to enable/disable interrupt on a
gpio.

Signed-off-by: 0hax <0hax@protonmail.com>

* uart: rename CONF_NR_BITS to CONF_DATA_BITS_NUMBER.

Signed-off-by: 0hax <0hax@protonmail.com>

* uart: use static const uint32_t instead of #define.

Signed-off-by: 0hax <0hax@protonmail.com>

* uart: use an enum to handle parity.

Signed-off-by: 0hax <0hax@protonmail.com>

* uart: split between esp32 and esp8266.

Signed-off-by: 0hax <0hax@protonmail.com>

* uart: check_uart_settings for parity and number of data bits.

Signed-off-by: 0hax <0hax@protonmail.com>

* name param data_bits

* add new params to test

Co-authored-by: Guillermo Ruffino <glm.net@gmail.com>

Commit:
Adding support for MAX31856 Thermocouple Temperature Sensor (feature #700) (#1039)

ahpohl committed:
Add support for additional Xiaomi BLE sensors (#1027) (#1027)

gitolicious committed:
Explicitly set language to English (#1073)

igg committed:
Added support for ssd1327 (#985)

Co-authored-by: Ilya Goldberg <iggie@mac.com>

Commit:
added energy reading for pzem004 (#1022)

rradar committed:
BH1750 Measurement time (#997)

* BH1750 Measurement time

Fixes https://github.com/esphome/feature-requests/issues/148

* lint

* add test

Co-authored-by: Otto Winter <otto@otto-winter.com>
Co-authored-by: Guillermo Ruffino <glm.net@gmail.com>

ivan4th committed:
Sort keys in dicts in output yaml for 'config' command (#1049)

Having different output each time due to random key order makes some
tasks harder (such as CI scripts).

ayufan committed:
Extend uart: with rx_buffer_size: (#1006)

* Extend `uart:` with `rx_buffer_size:`

This allows to configure `rx_buffer_size:`
to efficiently receive big payloads over UART

* lint

* remove old default value

* add test

Co-authored-by: Guillermo Ruffino <glm.net@gmail.com>

margau committed:
Add Webserver Prometheus support for sensor, binary sensor, fan, light, cover and switch (#1032)

apeeters committed:
Turn off PN532 RF field when not expecting a tag (#1046)

* Turn off PN532 RF field when not expecting a tag

Avoids interference with Wifi connectivity of nearby devices.

* Rename turn_off_rf_ method

* documented off command bytes

Co-authored-by: Guillermo Ruffino <glm.net@gmail.com>

igg committed:
making SPI CS optional (#988)

* making SPI CS optional

* CS pin should be declared as optional or required in CONFIG_SCHEMA

* changed SPI_DEVICE_SCHEMA to a function, like i2c

* added spi_device_schema() to pcd8544, lint fixes

* updated max31856 with new spi_device_schema()

* cleanup imports

Co-authored-by: Ilya Goldberg <iggie@mac.com>

freekode committed:
AQI calculator for HM3301 (#1011)

* HM3301 AQI calculator

* remove logs

* fixed after lint

* fixed after lint

* fixed after lint

* check NP for AQI sensor

* validation for AQI sensor

Skaronator committed:
RGBWW/CCT Lights: Fix gamma_correct when using constant_brightness (#1043)

ayufan committed:
Add E1.31 support (#950)

This adds a `e131` component that allows
to register `e131` addressable light effect.

This uses an internal implementation that
is thread-safe instead of using external libraries.

ayufan committed:
Add Adalight support (#956)

A component to support [Adalight](https://learn.adafruit.com/adalight-diy-ambient-tv-lighting).
This allows to control addressable LEDs over UART, by pushing data right into LEDs.

The most useful to use [Prismatik](https://github.com/psieg/Lightpack) to create
an immersive effect on PC.

Co-authored-by: Guillermo Ruffino <glm.net@gmail.com>

ayufan committed:
Add WLED support (#1092)

A component to support [WLED](https://github.com/Aircoookie/WLED/wiki/UDP-Realtime-Control).
This allows to control addressable LEDs over WiFi/UDP, by pushing data right into LEDs.

The most useful to use [Prismatik](https://github.com/psieg/Lightpack) to create
an immersive effect on PC.

It supports all WLED protocols:
- WARLS
- DRGB
- DRGBW
- DNRGB
- WLED Notifier

Co-authored-by: Guillermo Ruffino <glm.net@gmail.com>

Skaronator committed:
Update docker base image (#1093)

* Upgrade docker base image to 2.1.0

* Upgrade docker base image to 2.1.1

Co-authored-by: Otto Winter <otto@otto-winter.com>

ferbar committed:
SenseAir: flush input buffer on read error (#1017)

* SenseAir: flush input buffer on read error

* SenseAir: flush input buffer on read error - fix typo

* SenseAir: flush input buffer on read error - fix formating

* SenseAir: flush input buffer on read error - fix formating2

* SenseAir: flush input buffer on read error - fix formating3

* SenseAir: flush input buffer on read error - fix formating4

* Update esphome/components/senseair/senseair.cpp

* Update esphome/components/senseair/senseair.cpp

Co-authored-by: Guillermo Ruffino <glm.net@gmail.com>

blejdfist committed:
Add support for controlling fan direction (#1051)

* Fix fan oscillation trait not being used

* Add fan direction support to SpeedFan

* Add fan direction to API

* Add fan direction support to BinaryFan

* Fix CI errors

* Fix python format

* Change some ordering to trigger CI

* Add test for the configuration

JeffResc committed:
Add API component to logging error message (#1062)

* Add API component to logging error message

* lint

Co-authored-by: Guillermo Ruffino <glm.net@gmail.com>

SenexCrenshaw committed:
Add MCP3008 I/O Expander (#1057)

* Add MCP3008

* Fixed Travis build issues

* Travis Build Fix

* Fix Travis foramt issues

* Travis fixes

* Code review changes

* Fix bit shift. off by one and selecting wrong pin

* general fixes

* lint

Co-authored-by: Guillermo Ruffino <glm.net@gmail.com>

rnauber committed:
ADE7953: Fix dereferencing of a null pointer (#1086)

* ADE7953: Fix dereferencing of a null pointer

The ade7953 driver dereferences a null pointer, when not all of
its sensors are used. This gives an exception like:

Fatal exception:29 flag:2 (EXCEPTION) epc1:0x4020c241 epc2:0x00000000 epc3:0x00000000 excvaddr:0x00000018 depc:0x00000000

* Fix formatting

Co-authored-by: olg <x>

korellas committed:
sgp30 baseline write bug (#1157) (#1078)

SchumyHao committed:
use light remote value to trigger automation

Signed-off-by: SchumyHao <schumyhaojl@126.com>
